### ISSUE TITLE
Deadlock using ClientWebSocket

### ISSUE THREAD

**Issue by darxis** [Microsoft/dotnet#560]
.NET Framework 4.7.1

I have been experiencing deadlock situations using the `ClientWebSocket` class. My `ClientWebSocket` connects to a server that is not stable, and I need to reconnect to it several times per minute (I have no control over the WS server).

When reconnect happens, sometimes there is a deadlock situation when calling `ClientWebSocket .CloseAsync`. The control is not returned when calling `await webSocket.CloseAsync()`. I have created a dump file and inspected it with `System.pdb` symbols loaded. I can see that two threads are stuck deadlocked (stuck):
1. `WebSocketBase.StartOnCloseReceived` method at `Monitor.Enter(m_ThisLock, ref thisLockTaken);`
2. `WebSocketBase.CloseOutputAsyncCore` method at `TakeLocks(ref thisLockTaken, ref sessionHandleLockTaken);`

Something is not releasing the locks.

I have browsed through the .NET Framework 4.7.1 source and I think I might have found one bug. It could lie in the `WebSocketBase.CloseAsyncCore` method, which has a body:
```c#
private async Task CloseAsyncCore(WebSocketCloseStatus closeStatus,
            string statusDescription,
            CancellationToken cancellationToken)
        {
            ...something not relevant
 
            try
            {
                ThrowIfPendingException();
                if (IsStateTerminal(State))
                {
                    return;
                }
                ThrowIfDisposed();
 
                bool lockTaken = false;
                Monitor.Enter(m_ThisLock, ref lockTaken);
                ...something not relevant
            }
            finally
            {
                if (s_LoggingEnabled)
                {
                    Logging.Exit(Logging.WebSockets, this, Methods.CloseAsync, inputParameter);
                }
            }
        }
```

The problem is that in the `finally` clause there should be a call to `ReleaseLock(m_ThisLock, ref lockTaken);`, but there isn't. So if an exception happens, the lock is not released.

PS: When accessing the `ClientWebSocket` public methods I use a lock, so that only one thread is inside it (as it states in the documentation: "Any instance members are not guaranteed to be thread safe.")

**Comment by davidsh** [comment]
Thank you for the bug report.

We had reports from other customers about this potential deadlock.

Reference: Internal .NET Framework bug no:  497690.

We have fixed this bug and plan to release the fix as part of the .NET Framework 4.7.2 release planned for next year.

**Comment by darxis** [comment]
@davidsh Thank you for the fast response.

Which quarter of the 2018 is the 4.7.2 planned for? Is there any workaround until 4.7.2 released?

**Comment by davidsh** [comment]
>Which quarter of the 2018 is the 4.7.2 planned for?

At this point, I can only say H1, 2018.

>Is there any workaround until 4.7.2 released?

I'm not aware of any workaround since it is a race condition type of deadlock.

**Comment by darxis** [comment]
> I'm not aware of any workaround since it is a race condition type of deadlock.
Yes, but manually calling by reflection `ReleaseLocks` could unlock this. I am going to try and will give a feedback if it worked or not

**Comment by darxis** [comment]
I managed to `Monitor.Exit` two lock objects when the deadlock occurs: `m_SessionHandle` and `m_ThisLock`. I can see that the two Threads that were waiting no longer exists after exiting the monitors, but when calling `.Wait()` on the `CloseAsync` returned Task again, it still blocks, and I can see in the debugger that there are no other blocked tasks...

@davidsh Could you provide me with information about how exactly the bug was fixed/where was the problem? A simple `.patch` file/diff would do it.

**Comment by davidsh** [comment]
Unlike .NET Core which is open-source, .NET Framework is closed-source.  The updated sources will be published when .NET Framework 4.7.2 ships.  The current .NET Framework reference sources are located here:

https://github.com/Microsoft/referencesource

We are unable to provide specific closed-source changes for this bug fix at this time.

Similar to .NET Framework 4.7.1, .NET Framework 4.7.2 will have an Early Access program where you can install a pre-release version.

This was the information about the prior 4.7.1 Early Access:
https://blogs.msdn.microsoft.com/dotnet/2017/08/07/welcome-to-the-net-framework-4-7-1-early-access/

I am not sure of the exact date for the Early Access preview release of 4.7.2.

Also, GitHub CoreFx repo is not the normal place for reporting .NET Framework issues.  We encourage customers to report .NET Framework bugs/issues/suggestions here:

https://developercommunity.visualstudio.com/spaces/8/index.html

or

https://github.com/microsoft/dotnet

**Comment by letrungkien211** [comment]
Can somebody tell me if it fixes now? Will updating my project to 4.7.2 solve this issue?

**Comment by davidsh** [comment]
@karelz
